File: cpp/foxford/34/A.py
# n, m = map(int, input().split())
# a, b = map(int, input().split())

# adj = [[] for _ in range(n)]

# for _ in range(m):
#     f, t = map(int, input().split())
#     adj[f-1].append(t-1)
#     adj[t-1].append(f-1)

# used = [False] * n
# par = [-1] * n
# queue = [a-1]
# used[a-1] = True

# while len(queue) > 0:
#     v = queue.pop()

#     for u in adj[v]:
#         if not used[u]:
#             used[u] = True
#             par[u] = v
#             queue.append(u)

# if par[b-1] == -1:
#     print(-1)
#     exit(0)

# path = []
# j = b-1
# while j != -1:
#     path.append(j+1)
#     j = par[j]

# print(len(path))
# print(" ".join(map(str, reversed(path))))



from collections import deque, defaultdict

def bfs_shortest_path(edges, start, end):
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    queue = deque([(start, [start])])
    visited = set([start])
    
    while queue:
        current, path = queue.popleft()
        if current == end:
            return len(path) - 1, path
        for neighbor in graph[current]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append((neighbor, path + [neighbor]))
                
    return -1

n, m = map(int, input().split())
start, end = map(int, input().split())

edges = []
for _ in range(m):
    u, v = map(int, input().split())
    edges.append((u, v))

result = bfs_shortest_path(edges, start, end)

if result == -1:
    print(-1)
else:
    length, path = result
    print(length)
    print(" ".join(map(str, path)))

File: cpp/foxford/34/A.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>

using namespace std;

vector<int> bfs_shortest_path(const unordered_map<int, vector<int>> &graph, int start, int end)
{
    queue<pair<int, vector<int>>> q;
    unordered_set<int> visited;
    q.push({start, {start}});
    visited.insert(start);

    while (!q.empty())
    {
        auto pair = q.front();
        auto current = pair.first;
        auto path = pair.second;
        q.pop();

        if (current == end)
        {
            return path;
        }

        for (int neighbor : graph.at(current))
        {
            if (visited.find(neighbor) == visited.end())
            {
                visited.insert(neighbor);
                vector<int> new_path = path;
                new_path.push_back(neighbor);
                q.push({neighbor, new_path});
            }
        }
    }
    return {};
}

int main()
{
    int n, m;
    cin >> n >> m;
    int start, end;
    cin >> start >> end;

    unordered_map<int, vector<int>> graph;
    for (int i = 0; i < m; ++i)
    {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    vector<int> result = bfs_shortest_path(graph, start, end);

    if (result.empty())
    {
        cout << -1 << endl;
    }
    else
    {
        cout << result.size() - 1 << endl;
        for (int node : result)
        {
            cout << node << " ";
        }
        cout << endl;
    }

    return 0;
}
